pkg/controllers/updaterun/execution_test.go

```diff
@@ -398,3 +407,522 @@ func TestBuildApprovalRequestObject(t *testing.T) {
+						State:                placementv1beta1.BindingStateScheduled, // But not Bound yet.
+					},
+				},
+			},
+			interceptorFunc: &interceptor.Funcs{
+				Update: func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.UpdateOption) error {
+					return errors.New("failed to update binding state")
+				},
+			},
+			wantErr:      errors.New("failed to update binding state"),
+			wantWaitTime: 0,
+		},
+		{
+			name: "binding synced and bound but generation updated - update rolloutStarted fails",
+			updateRun: &placementv1beta1.ClusterStagedUpdateRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "test-update-run",
+					Generation: 1,
+				},
+				Spec: placementv1beta1.UpdateRunSpec{
+					PlacementName:         "test-placement",
+					ResourceSnapshotIndex: "1",
+				},
+				Status: placementv1beta1.UpdateRunStatus{
+					StagesStatus: []placementv1beta1.StageUpdatingStatus{
+						{
+							StageName: "test-stage",
+							Clusters: []placementv1beta1.ClusterUpdatingStatus{
+								{
+									ClusterName: "cluster-1",
+									// No conditions - cluster has not started updating yet.
+								},
+							},
+						},
+					},
+					UpdateStrategySnapshot: &placementv1beta1.UpdateStrategySpec{
+						Stages: []placementv1beta1.StageConfig{
+							{
+								Name:           "test-stage",
+								MaxConcurrency: &intstr.IntOrString{Type: intstr.Int, IntVal: 1},
+							},
+						},
+					},
+				},
+			},
+			bindings: []placementv1beta1.BindingObj{
+				&placementv1beta1.ClusterResourceBinding{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "binding-1",
+						Generation: 2, // Generation updated by scheduler.
+					},
+					Spec: placementv1beta1.ResourceBindingSpec{
+						TargetCluster:        "cluster-1",
+						ResourceSnapshotName: "test-placement-1-snapshot",        // Already synced.
+						State:                placementv1beta1.BindingStateBound, // Already Bound.
+					},
+					Status: placementv1beta1.ResourceBindingStatus{
+						Conditions: []metav1.Condition{
+							{
+								Type:               string(placementv1beta1.ResourceBindingRolloutStarted),
+								Status:             metav1.ConditionTrue,
+								ObservedGeneration: 1, // Old generation - needs update.
+								Reason:             condition.RolloutStartedReason,
+							},
+						},
+					},
+				},
+			},
+			interceptorFunc: &interceptor.Funcs{
+				SubResourceUpdate: func(ctx context.Context, client client.Client, subResourceName string, obj client.Object, opts ...client.SubResourceUpdateOption) error {
+					// Fail the status update for rolloutStarted.
+					return errors.New("failed to update binding rolloutStarted status")
+				},
+			},
+			wantErr:      errors.New("failed to update binding rolloutStarted status"),
+			wantWaitTime: 0,
+		},
+		{
+			name: "binding synced, bound, rolloutStarted true, but binding has failed condition",
```

Comment:
Can you merge the binding related test cases into one? To test the aggregateErr effect?